server/service/client.go

```diff
@@ -303,45 +303,80 @@ func (c *Client) runAppConfigChecks(fn func(ac *fleet.EnrichedAppConfig) error)
-
-	for _, profile := range profiles {
-		filePath := resolveApplyRelativePath(baseDir, profile.Path)
-		fileContents, err := os.ReadFile(filePath)
-		if err != nil {
-			return nil, fmt.Errorf("applying fleet config: %w", err)
-		}
-
-		if expandEnv {
-			fileContents, err = spec.ExpandEnvBytes(fileContents)
+func getProfilesContents(baseDir string, macProfiles []fleet.MDMProfileSpec, windowsProfiles []fleet.MDMProfileSpec, expandEnv bool) ([]fleet.MDMProfileBatchPayload, error) {
+	// map to check for duplicate names across all profiles
+	extByName := make(map[string]string, len(macProfiles))
+	result := make([]fleet.MDMProfileBatchPayload, 0, len(macProfiles))
+
+	// iterate over the profiles for each platform
+	for platform, profiles := range map[string][]fleet.MDMProfileSpec{
+		"macos":   macProfiles,
+		"windows": windowsProfiles,
+	} {
+		for _, profile := range profiles {
+			filePath := resolveApplyRelativePath(baseDir, profile.Path)
+			fileContents, err := os.ReadFile(filePath)
 			if err != nil {
-				return nil, fmt.Errorf("expanding environment on file %q: %w", profile.Path, err)
+				return nil, fmt.Errorf("applying custom settings: %w", err)
 			}
-		}
 
-		// by default, use the file name. macOS profiles use their PayloadDisplayName
-		ext := filepath.Ext(filePath)
-		name := strings.TrimSuffix(filepath.Base(filePath), ext)
-		if mdm.GetRawProfilePlatform(fileContents) == "darwin" && ext == ".mobileconfig" {
-			mc, err := fleet.NewMDMAppleConfigProfile(fileContents, nil)
-			if err != nil {
-				return nil, fmt.Errorf("applying fleet config: %w", err)
+			if expandEnv {
+				fileContents, err = spec.ExpandEnvBytes(fileContents)
+				if err != nil {
+					return nil, fmt.Errorf("expanding environment on file %q: %w", profile.Path, err)
+				}
 			}
-			name = strings.TrimSpace(mc.Name)
-		}
-		if e, isDuplicate := extByName[name]; isDuplicate {
-			return nil, errors.New(fmtDuplicateNameErrMsg(name, e, ext))
-		}
-		extByName[name] = ext
-		result = append(result, fleet.MDMProfileBatchPayload{
-			Name:     name,
-			Contents: fileContents,
-			Labels:   profile.Labels,
-		})
 
+			ext := filepath.Ext(filePath)
+			// by default, use the file name (for macOS mobileconfig profiles, we'll switch to
+			// their PayloadDisplayName when we parse the profile below)
+			name := strings.TrimSuffix(filepath.Base(filePath), ext)
+			// for validation errors, we want to include the platform and file name in the error message
+			prefixErrMsg := fmt.Sprintf("Couldn't edit %s_settings.custom_settings (%s%s)", platform, name, ext)
+
+			// validate macOS profiles
+			if platform == "macos" {
+				switch ext {
+				case ".mobileconfig", ".xml": // allowing .xml for backwards compatibility
+					mc, err := fleet.NewMDMAppleConfigProfile(fileContents, nil)
+					if err != nil {
+						return nil, fmt.Errorf("%s: %s", prefixErrMsg, strings.TrimPrefix(err.Error(), "new MDMAppleConfigProfile: "))
```

Comment:
I understand the reason for this, but this will be brittle, a seemingly innocuous change in `NewMDMAppleConfigProfile` can break this error message. I think it could be a bit more robust by just unwrapping the error once, if we know that it is always wrapped with the extra message prefix? I.e. something like `errForMsg := err.Unwrap(); if errForMsg == nil { errForMsg = err }`, and then you wouldn't need to trim the prefix?

Reply:
That unwrapping suggestion is really neat, but I don't think we can do `Unwrap` here because we aren't wrapping with `ctxerr` or anything that supports that method upstream. I suppose we could use it upstream anyway with a background context, but I think we have generally tried to avoid that. Is there another wrapping pattern we should consider here?

In this case, I tried to be mindful of future changes potentially breaking things and thought that worst case here we'd just end up with an uglier (more verbose) error message (TrimPrefix has nothing to trim and would return the original message unaltered). I can add a quick test to surface any changes in how the upstream method returns errors if we want to be extra careful.

Reply:
It looks like `NewMDMAppleConfigProfile` uses `fmt.Errorf` with a `%w` directive, so we could pass
the error from that function to `errors.Unwrap` here. Could that be a solution? 

https://pkg.go.dev/fmt#Errorf

Reply:
Hmm, how does the caller access that method? Does we need to cast the returned error as something to check if it implements `Unwrap`? Is there a general `Unwrapper` interface defined somewhere?

Reply:
Doesn't look like we need a cast. See this Go playground, where I modified the example program from the docs: https://go.dev/play/p/SEfzgWajhOs

Reply:
heads-up: this is what `errors.Unwrap` does, that's why Martin is suggesting the nil check

```
// Unwrap returns the result of calling the Unwrap method on err, if err's
// type contains an Unwrap method returning error.
// Otherwise, Unwrap returns nil.
//
// Unwrap only calls a method of the form "Unwrap() error".
// In particular Unwrap does not unwrap errors returned by [Join].
func Unwrap(err error) error {
	u, ok := err.(interface {
		Unwrap() error
	})
	if !ok {
		return nil
	}
	return u.Unwrap()
}
```

Reply:
Ah, ok. Martin's suggestion was `err.Unwrap()` rather than `errors.Unwrap(err)`.

Reply:
Sorry for the confusion, `err.Unwrap` was a typo, I meant `errors.Unwrap` indeed.